cpp/src/barretenberg/honk/proof_system/work_queue.hpp

```diff
@@ -0,0 +1,143 @@
+#pragma once
+
+#include "barretenberg/honk/transcript/transcript.hpp"
+#include "barretenberg/plonk/proof_system/proving_key/proving_key.hpp"
+#include <cstddef>
+#include <memory>
+
+namespace proof_system::honk {
+
+// Currently only one type of work queue operation but there will likely be others related to Sumcheck
+enum WorkType { SCALAR_MULTIPLICATION };
+
+template <typename Params> class work_queue {
+
+    using CommitmentKey = typename Params::CK;
+    using FF = typename Params::Fr;
+    using Commitment = typename Params::C;
+
+    struct work_item_info {
+        uint32_t num_scalar_multiplications;
+    };
+
+    struct work_item {
+        WorkType work_type = SCALAR_MULTIPLICATION;
+        std::span<FF> mul_scalars;
+        std::string label;
+    };
+
+  private:
+    std::shared_ptr<proof_system::plonk::proving_key> key;
+    // TODO(luke): Consider handling all transcript interactions in the prover rather than embedding them in the queue.
+    proof_system::honk::ProverTranscript<FF>& transcript;
+    CommitmentKey commitment_key;
+    std::vector<work_item> work_item_queue;
+
+  public:
+    explicit work_queue(std::shared_ptr<proof_system::plonk::proving_key>& proving_key,
+                        proof_system::honk::ProverTranscript<FF>& prover_transcript)
+        : key(proving_key)
+        , transcript(prover_transcript)
+        , commitment_key(key->circuit_size, "../srs_db/ignition"){}; // TODO(luke): make this properly parameterized
+
+    work_queue(const work_queue& other) = default;
+    work_queue(work_queue&& other) noexcept = default;
+    work_queue& operator=(const work_queue& other) = delete;
+    work_queue& operator=(work_queue&& other) = delete;
+    ~work_queue() = default;
+
+    [[nodiscard]] work_item_info get_queued_work_item_info() const
+    {
+        uint32_t scalar_mul_count = 0;
+        for (const auto& item : work_item_queue) {
+            if (item.work_type == WorkType::SCALAR_MULTIPLICATION) {
+                ++scalar_mul_count;
+            }
+        }
+        return work_item_info{ scalar_mul_count };
+    };
+
+    [[nodiscard]] FF* get_scalar_multiplication_data(size_t work_item_number) const
+    {
+        size_t count = 0;
+        for (const auto& item : work_item_queue) {
+            if (item.work_type == WorkType::SCALAR_MULTIPLICATION) {
+                if (count == work_item_number) {
+                    return const_cast<FF*>(item.mul_scalars.data());
```

Comment:
Why is this needed? I think it's easy to get undefined behavior when using `const_cast`. Can't we just return the span?

Reply:
Yeah I agree but two things: 1) this function is only used in the WASM/c_bind context where it expects an `fr*` so I didn't want to mess with it too much. 2) It may actually be necessary (at least for now) because the pedersen implementation makes heavy use of pointer arithmetic. (You'll notice that the CommitmentKey is also extracting a raw pointer from a span behind the scenes before passing to perdersen). I'm not sure how hard it would be to update pedersen to use a span somehow but needless to say this is not the time/place for that.

Reply:
Fair enough, but we should keep this in mind.